[RNMobile] Add Featured Banner to Image Block (Android Only)

SiobhyB opened the pull request:
Paritial fix for: wordpress-mobile/gutenberg-mobile#1011

`gutenberg-mobile`: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3371
`WordPress-Android`: https://github.com/wordpress-mobile/WordPress-Android/pull/14451

## Description

This PR introduces a **Featured** banner, which overlays a post's featured image when it's added to an image block. This will make it easier for users to identify which of the images in their posts, if any, are set as featured directly from the editor.

_Note: This PR introduces the **Featured** banner on Android devices only, support for iOS will follow in future iterations._

## How has this been tested?

### Test Case 1: General Flow

To start with, an image block should be updated to reflect whether its image is featured or not when set via the general Post Settings flow:

* Tap the three icons to the upper right of the post editor to access **Post Settings** and set a featured image.
* Tap the back button. If the new featured image hasn't already been added to the editor via an image block, add it.  
* Confirm that the **Featured** banner displays overlays the correct image block.
* Return to **Post Settings**. Tap the existing featured image and then **Remove**.
* Tap the back button to confirm that the **Featured** banner has been removed from the image block, as expected.

### Test Case 2: "Editor Loads" Flow

Next, we should verify that a **Featured** banner appears when a post with a featured image is first loaded. To test this, follow these steps:

* Verify that a featured image has been set on the post and tap the back button.
* Return to the post's editor by tapping on it from the post list screen.
* Verify that the post's featured image has a **Featured** banner overlaying it.

### Test Case 3: "Replace in Image Block" Flow

Finally, it's possible to replace an image directly within the image block's settings. We need to verify that the **Featured** banner updates as expected depending on the image that's chosen for an image block:

* Select any image block that is flagged with the **Featured** banner.
* Tap the edit icon to the upper right and then **Replace**. 
* Replace with any image that _isn't_ the post's featured image.
* Verify that the **Featured** banner is removed from the block when the new image uploads.
* Go through the steps to replace the block's image again, this time pick the post's featured image.
* Verify that the **Featured** banner appears again.

## Screenshots <!-- if applicable -->

| Before  | After |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/114570411-b7907c80-9c6d-11eb-9d3c-8672b0771d0b.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/114570332-a5164300-9c6d-11eb-88f7-fe751259087e.png" width="100%"> |

## Types of changes

This PR introduces a new feature (a non-breaking change that adds functionality). Specifically, it adds a **Featured** banner that overlays an image block containing a post's featured image. The code can be split into two overarching categories:

### Badge Component

* A [new Badge component](https://github.com/WordPress/gutenberg/pull/30806/files#diff-f3d1866bb6de6b07c40f6cc1e97d08ef289dc2b4874deec8c414541e0dd5844e) has been created, which adds a "badge" with some text to its child component's upper left. 
* In this case, this new component is used to add a **Featured** badge to an image block. 
* The component was created with re-usability in mind and [the README](https://github.com/WordPress/gutenberg/pull/30806/files#diff-2500f619901448109c703529708b09a157df9072a96e99c19d1d217d54e9aa24) describes how it may be used in other areas of the codebase.

### Logic for Fetching Featured Image ID

* To fetch the ID of a post's initial featured image, [a prop is passed over the bridge](https://github.com/WordPress/gutenberg/pull/30806/files#diff-906aa52029680055db7dc413437ac227ebeaef57c0da27347a019fc53506099a) and set in [the post object](https://github.com/WordPress/gutenberg/pull/30806/files#diff-d45b169f7a1824d1f08790b97680e012bb30be66ed5714994a3959ba23cdee94R146).
* An [event subscription](https://github.com/WordPress/gutenberg/pull/30806/files#diff-bf4c994cef3006848617e4cd8a8ad4bfbc79ef6feda4d58152369ec92019d51e) is used to track a change to a post's featured image from the Android app. The `editEntityRecord` selector is then used in combination with listeners to [update the value stored for a post's featured image ID](https://github.com/WordPress/gutenberg/pull/30806/files#diff-d45b169f7a1824d1f08790b97680e012bb30be66ed5714994a3959ba23cdee94). 
* The above enables a `featuredImageId` to be passed to and [used in the image component's edit.native.js file](https://github.com/WordPress/gutenberg/pull/30806/files#diff-7126ae03000a022bcefa0d2a5fb03b6aca56ed4047e3a49d2131babf229cf6c1). In order to determine whether or not to display the badge component, the `featuredImageId` is compared to the ID of the image in the block.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

SiobhyB committed:
Add "set as featured image" button to image block

This commit adds a "set as featured image" button to the image block's settings and also styles it.

SiobhyB committed:
Add "setFeaturedImage" function in bridge code

With this commit, a "setFeaturedImage" function has been added to the React Native bridge code.

SiobhyB committed:
Add setFeaturedImage function to MainApplication

This commit adds the setFeaturedImage function to the MainApplication file, so that it works as expected with the demo apps.

SiobhyB committed:
Add "OnSetFeaturedImageListener" listener

This commit sets up a listener called "OnSetFeaturedImageListener" in order to allow communication with the native apps.

SiobhyB committed:
Update edit.native.js with small syntax correction

This commit updates edit.native.js with small syntax correction, there was a missing comma at the end of a statement.

SiobhyB committed:
Set up "OnSetFeaturedImageListener" in WPAndroidGlue.java

This commit updates WPAndroidGlue.java in order to set up a "OnSetFeaturedImageListener" function that will communicate with the Android app when a user clicks a "set as featured image" button.

SiobhyB committed:
Passing featured image ID from native (Android) to Gutenberg

This commit seeks to send information about a post's featured image ID to Gutenberg by utlising functions from Android.

SiobhyB committed:
Merge branch 'trunk' into update/featured-image-button-in-image-block-settings

SiobhyB committed:
Restore missing curly brace

Accidentally removed a curly brace with the last commit, so restoring it with this one.

SiobhyB committed:
Updates to featured-banner.native.js

This commit including updates to featured-banner.native.js, a component for the "featured" banner that will overlay the image block when an image is featured. "payload.featuredImageId" logs the ID of a featured image after it's updated.

SiobhyB committed:
Attempts to select featured image using getEditedPostAttribute()

This commit includes attempts to select a post's featured image using getEditedPostAttribute().

SiobhyB committed:
Update featured-banner.native.js

With this commit, I'm updating featured-banner.native.js with the latest changes to that component's UI.

SiobhyB committed:
Add featured banner to image/index.native.js

With this commit, a featured banner is added to the image component.

SiobhyB committed:
Update featured image banner and button styles

This commit updates the styles for the featured image banner and button that overlay the image compoonent.

SiobhyB committed:
Create getFeaturedImageId function and cleanup

This commit includes a new getFeaturedImageId function, for grabbing a post's featured image when a component is mounted and also cleans up some unused code.

SiobhyB committed:
Tidy up edit.native.js

SiobhyB committed:
Update AndroidGlueCode

Update AndroidGlueCode to include OnFocalPointPickerTooltipShownEventListener.

SiobhyB committed:
retrigger checks

SiobhyB committed:
Update setup.js

SiobhyB committed:
retrigger checks

SiobhyB committed:
Update function name from "featuredImageIdChange" to "featuredImageIdCurrent"

With this commit, I'm updating a function's name from "featuredImageIdChange" to "featuredImageIdCurrent" in order to more accurately reflect its purpose.

SiobhyB committed:
Rename "featuredImageIdNotifier" to "onRequestFeaturedImageId"

This commit renames "featuredImageIdNotifier" to "onRequestFeaturedImageId" in order to clarify the function's purpose.

SiobhyB committed:
Create OnSetFeaturedImageListener

This commit introduces an "OnSetFeaturedImageListener" and moves new featured image related functions to it from "OnMediaLibraryButtonListener".

SiobhyB committed:
Dismiss bottomsheet when "setFeaturedImage" is called

SiobhyB committed:
Introduce "onRemoveFeatured" button and tidying up names

This commit introduces a "onRemoveFeatured" button that displays in the image block's settings when an image is featured. It also changes the name of the "onGetFeaturedImageId" function to "checkIfFeaturedImage", to better reflect its current functionality.

SiobhyB committed:
Updates to function names

This commit updates "onRequestFeaturedImageId" to "sendToJSFeaturedImageId" and "getFeaturedImageId" to "checkIfFeaturedImage". In both cases, this has been done to clarify the purpose of the functions.

SiobhyB committed:
Convert FeaturedBanner to Badge component

This commit converts the FeaturedBanner component to a more generic Badge component. This is to make it more re-usable. There are also some small styling tweaks included in this commit, including changes to the component's border-radius.

SiobhyB committed:
Update edit.native.js to correct merge conflict

SiobhyB committed:
Correct merge conflict with GutenbergBridgeJS2Parent.java

This commit is an attempt to correct this PR's merge conflict with GutenbergBridgeJS2Parent.java.

SiobhyB committed:
retrigger checks

SiobhyB committed:
retrigger checks

SiobhyB committed:
retrigger checks

SiobhyB committed:
retrigger checks

SiobhyB committed:
Update edit.native.js to fix merge conflict

SiobhyB committed:
Update edit.native.js to check if an image is featured uppn changes

This commit fixes a bug, where the "featured" banner did not update if an image was replaced directly within a block, using the available settings to the upper right.

SiobhyB committed:
Update styles.native.scss to add border to the top of featured button

This commit also tidies up some of the code around styling to account for dark mode.

SiobhyB committed:
Update edit.native.js

This commit moves the "checkIfFeaturedImage" check into its own function.

SiobhyB committed:
Introduce "featuredMedia" prop for use when editor loads

This commit introduces a "featuredMedia" prop that is intended to be used when Gutenberg first loads. The prop is going to be passed over the bridge so that we can get information about a post's featured image ID from the app.

SiobhyB committed:
Fetch ID of post's initial featured image

This commit introduces some changes in react-native-editor/src/index.js that enable getEditedPostAttribute( 'featured_media' ) to fetch the initial ID of a post's featured image. This is then used in the image block's componentDidMount function to pinpoint a featured image when it's first loaded in the editor.

SiobhyB committed:
Merge branch 'trunk' into update/featured-image-button-in-image-block-settings

SiobhyB committed:
Updated featured state when image is replaced within block

This commit uses "setAttributes" to define a "currentFeaturedImageId" attribute. This attribute is then used in "componentDidUpdate" to check if an image is featured if it's replaced directly within the image block. There are also some smaller tweaks/improvements to the code in this commit, such as changing an if/else statement to use ES6's shorthand.

SiobhyB committed:
Remove checkIfFeaturedImage function

This function has been replaced with functionality to grab a post's initial ID, as can be seen in the previous couple of commits.

SiobhyB committed:
Fix merge conflict

SiobhyB committed:
Fix merge conflict

SiobhyB committed:
Add "setFeaturedImage" and "featuredImageIdCurrent" functions to Swift files

SiobhyB committed:
Simplify this.state.isFeaturedImage by converting to a boolean insider render()

This commit simplifies the process of determining whether an image is featured (currently done via by setting an isFeaturedImage state) by converting it to a boolean insider the render() function.

SiobhyB committed:
Comment out iOS-specific changes to the bridge

This commit comments out iOS-specific changes to the bridge in ordet to test whether these changes are the reason behind a failing performance test.

SiobhyB committed:
Add iOS functions to bridge

SiobhyB committed:
Flag featured image when editor mounts and comment out iOS methods from bridge

This commit adds a check to "componentDidMount" so that a featured image is flagged when the editor first mounts (fixes previous regression). It also comments out iOS methods over the bridge, a temporary change.

SiobhyB committed:
Comment out a bridge function for iOS (missing from previous commit).

SiobhyB committed:
Add set featured image functionality behind dev flag

SiobhyB committed:
Move "set as featured" button behind devOnly and androidOnly flag

This commit also changes the "featured" banner so that it's accessible to all.

SiobhyB committed:
Remove redundant code

This PR branches off #28854 in order to simplify PRs for a new feature that makes it easier for users to set a featured image from the image block. With this commit, I'm removing redundant code that isn't necessary for this specific PR (which will only include code that adds a "featured" banner to the image block).

SiobhyB committed:
Remove import of android.util.Log

SiobhyB committed:
Update CHANGELOG.md

hypest commented:
Thanks for separate PR just to cover the "display featured banner" functionality @SiobhyB , I will take a look tomorrow!

SiobhyB committed:
Update Badge component for re-usability

This commit updates the Badge component so that it works by wrapping around another component. This is designed to make the component easier to re-use.

SiobhyB committed:
Merge branch 'add-featured-badge' of https://github.com/SiobhyB/gutenberg into add-featured-badge

SiobhyB committed:
merge

SiobhyB committed:
Correct merge conflict

hypest commented:
Testing results:
✅ Test Case 1: General Flow
✅ Test Case 2: "Editor Loads" Flow
✅ Test Case 3: "Replace in Image Block" Flow

SiobhyB committed:
Lift subscribeFeaturedImageIdCurrent from image/edit.native.js to post-edit/edit.native.js

This components lifts the subscribeFeaturedImageIdCurrent function so that the data isn't managed directly from within the image component.

SiobhyB committed:
Remove commented out functions from iOS bridge

This commit also removes a redundant empty space in a bridge file.

hypest commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -153,13 +151,6 @@ export class ImageEdit extends Component {
-
-		// Flag the ID of a post's featured image when editor mounts.
-		if ( attributes.id ) {
-			setAttributes( { featuredImageId } );
```
Good to see this removal Siobhan! I've been meaning to comment on the PR for it, in the sense that the Image component shouldn't keep `featuredImageId` in its attribute list as it's not an attribute of the block itself but rather, a property of the global post object. 👍

SiobhyB replied:
Thank you!

hypest commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -445,6 +447,11 @@ export class ImageEdit extends Component {
+		const isFeaturedImage = featuredImageId === attributes.id;
+
+		// eslint-disable-next-line no-unused-vars
+		const androidOnly = Platform.OS === 'android';
```
Looks like the `androidOnly` const is not used anywhere, can we remove it?

SiobhyB replied:
Ah, good point! I tested and there aren't any issues loading posts in iOS with this PR, so the androidOnly tag seems safe to remove. I've removed that now.

hypest commented on packages/components/src/mobile/badge/README.md:
```diff
@@ -0,0 +1,31 @@
+# Badge
+
+The Badge component is designed to be wrapped around another component. It add a "badge" with some text to the child component's upper left.
```
> add

Typo perhaps? Replace with "adds"?

SiobhyB replied:
A typo indeed, thank you for spotting that! I've updated.

hypest commented on packages/edit-post/src/editor.native.js:
```diff
@@ -173,8 +193,10 @@ export default compose( [
 	} ),
 	withDispatch( ( dispatch ) => {
 		const { switchEditorMode } = dispatch( editPostStore );
+		const { editEntityRecord } = dispatch( 'core' );
```
I think using the store name directly is not preferred anymore (see [this PR that recently refactored lots of code to switch to using the exposed name](https://github.com/WordPress/gutenberg/pull/27414) instead). Let's follow the same pattern as the `editPostStore` above, thanks!

SiobhyB replied:
Ah, thank you for raising that and the additional context, I've gone ahead to update it. 🙇‍♀️

SiobhyB committed:
Remove androidOnly const

The androidOnly const is not in use and can be removed.

SiobhyB committed:
Correct typo in README

"add" should have been "adds". This commit corrects that.

SiobhyB committed:
Replace call to 'core' with 'coreStore"

This commit replaces the direct call to the 'core' store with a call to 'coreStore', in line with preferred patterns followed elsewhere in the project.

hypest commented on packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/WPAndroidGlue/GutenbergProps.kt:
```diff
@@ -12,6 +12,7 @@ data class GutenbergProps @JvmOverloads constructor(
     val enableAudioBlock: Boolean,
     val localeSlug: String,
     val postType: String,
+    val featuredMedia: Int,
```
Kinda a nit but, wdyt about just using `featuredImageId` pretty much everywhere, except the Post object itself where the property is named `featuredMedia`? Both the apps and Gutenberg-desktop refer to "featured image" on the UI and code. I feel like using the `Media` part in places just adds some confusion. Thoughts?

SiobhyB replied:
I'd had this on my mind a little too, I used `featuredMedia` as it was a bit more inline with `getEditedPostAttribute('featured_media')`, but agree that using `featuredImageId` seems clearer overall. I've updated this PR and will update the Android PR too when the tests have run.

hypest replied:
Should also add that I'm not sure if WordPress would accept a _video_ as featured on a post, in which case "featured media" is a more appropriate name. Gutenberg-desktop doesn't seem to offer videos when selecting a featured media though.

SiobhyB committed:
Rename featuredMedia to featuredImageId

hypest commented on packages/edit-post/src/editor.native.js:
```diff
@@ -80,6 +80,8 @@ class Editor extends Component {
 	}
 
 	componentDidMount() {
+		const { editEntityRecord, postType, postId } = this.props;
```
Friendly reminder to resist the urge to have different type of changes in the same commit, unless they form some union of sorts. In this case (commit 427845acbbeb52f071fcf56f8225228106dcb1e8), cleaning the "androidOnly" code and using destructuring for those props are not a union so I'd put them in separate commits.

In many cases, our local file changes probably include all the changes, and what I use `for putting them in separate commits is `git add -p` to interactively select which individual changes to stage. [This article seems to have some good material on this](https://johnkary.net/blog/git-add-p-the-most-powerful-git-feature-youre-not-using-yet/). Other (GUI?) git clients have the feature as well, usually named as simply "staging specific lines".

SiobhyB replied:
Thank you for your insight here and for emphasising the importance of separate, clear commits. I've gone ahead to separate these two out and will bear this in mind for the future.

hypest commented on packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/WPAndroidGlue/DeferredEventEmitter.java:
```diff
@@ -40,6 +42,8 @@
     private static final String EVENT_NAME_MEDIA_SAVE = "mediaSave";
     private static final String EVENT_NAME_MEDIA_REPLACE_BLOCK = "replaceBlock";
 
+    private static final String EVENT_FEATURED_IMAGE_ID_CURRENT = "featuredImageIdCurrent";
```
(Another nit?) Now that on the JS side we use the core-data Store for featured image id storage, I can't quite be sure what we mean when we say "Current". I know that [it was me not long before](https://github.com/WordPress/gutenberg/pull/28854#discussion_r598733588) that proposed the name, but back then I felt that the source of truth is the data we keep in the native side. That's kinda still true but when we'll add the functionality of setting the featured image from the Image block, the source of truth will the be image block itself or the updated core-data store if we go that way. In that case, what's "current"?

Long story short, wdyt about renaming to `featuredImageIdNativeUpdated`? This will also help on the JS side where reading about subscribing to such an event it's clear that it's data held on the native side.

SiobhyB replied:
Your reasoning behind the name change makes sense to me here, I've gone ahead to make that update in this PR and the Android update will follow soon.

SiobhyB committed:
Destructure props

This commit destructures props in the componentDidMount function to improve readability

SiobhyB committed:
Rename "featuredImageIdCurrent" to "featuredImageIdNativeUpdated"

Following discussion, this commit renames "featuredImageIdCurrent" to "featuredImageIdNativeUpdated" in order to more clearly describe the function's purpose.

hypest reviewed (approved):
The Gutenberg side PR looks good to me at this stage @SiobhyB , good job and thank you for all your follow ups to my review feedback! ❤️

I understand that the WPAndroid side PR is still under review so let's hold on merging the Gutenberg side until all PRs are ready and green.

In the meantime though, may I suggest updating the PR description to match the latest of the code changes? In particular, I'd suggest expanding the "Types of changes" section with details of the various qualitatively different changes in the PR. For example, the introduction of the Badge component, using an event subscription for the featured id change from native and so on. Ideally, each bullet would have some small explanation of the use or need of the change (especially if it's non trivial to understand the "why"), plus a link to the diff or commit so the reader can dig in if they want. An example of such an expanded description could be [this PR I opened recently](https://github.com/WordPress/gutenberg/pull/29683). Let me know if anything is not clear, thanks!

hypest commented:
Come to think of it again, the bridge changes might actually break the WPiOS app since we're only adding the related support on WPAndroid's side, right? Can you check @SiobhyB ? Thanks!

SiobhyB committed:
Limit subscribeFeaturedImageIdNativeUpdated to Android-only

subscribeFeaturedImageIdNativeUpdated is currently only supported in the Android app, so this commit limits its usage to prevent errors on iOS.

SiobhyB commented:
@hypest, thank you for this! A fix has been committed now. ✅ I'm still working on cleaning up commit history for the other two repositories, also. I'll comment on the Android issue when I'm done there.

_Edited to add: I've also added to the PR, as per your guidance. Let me know if there are any further improvements you see. :)_

hypest commented:
> Edited to add: I've also added to the PR, as per your guidance. Let me know if there are any further improvements you see. :)

Wow, looks awesome to me, thanks! Let me know if you find this part verbose perhaps or not, curious about your thoughts after adding the info.

hypest commented:
In the meantime, I verified the crash on open (screenshot shared below) and the https://github.com/WordPress/gutenberg/pull/30806/commits/f520659cc578757f9e800b8e308c2a783175a7aa fix! 🎉 

<img src="https://user-images.githubusercontent.com/1032913/114912091-45b06280-9e28-11eb-9636-6401a6f13330.jpg" width="300px">

SiobhyB commented:
> Wow, looks awesome to me, thanks! Let me know if you find this part verbose perhaps or not, curious about your thoughts after adding the info.

Thank you!

I liked creating a summary of the changes. It's probably something I'll try to do "as I go" in the future, as it seems it might be a handy "tool" for making sure the code makes sense to me and a good way to spot any room for improvement early on. 

I could have likely been more succinct in this, but I think the wordiness is a product of my newness to code and not having an instinct for the most important parts. I'm sure that'll change given time. :)

hypest commented:
> I could have likely been more succinct in this, but I think the wordiness is a product of my newness to code and not having an instinct for the most important parts. I'm sure that'll change given time. :)

Just to be clear, it doesn't read verbose to me. But I was interested to hear how did you find the experience of crafting that sections and whether at any point felt too "bureaucratic" or verbose perhaps.

hypest reviewed (approved):
LGTM @SiobhyB !

Here's what I think can be the plan now:
0. I will LGTM all three PRs (Gutenberg, gutenberg-mobile, WPAndroid)
1. I will merge the Gutenberg PR 🎉
2. Please then, update the gutenberg-mobile PR to point to the newly merged hash on Gutenberg's trunk
3. Wait for CI to get green in gutenberg-mobile. Feel free to use the "automatically merge on green" functionality.
4. Please get in touch with the native block editor release wranglers to ensure the two processes (release and this PRs merging) align well.
4. Update the WPAndroid PR to point to the a properly tagged gb-mobile pointer. 
5. After CI is green on WPAndroid, merge it!
6. Celebrate! 😄

SiobhyB commented:
Thank you @hypest! Just noting here for posterity that, after discussing with the release wranglers, we'll be waiting until next week to merge these changes.

SiobhyB commented:
@hypest, would you be able to merge this today? I'll then go ahead with the other steps you listed. :) Thank you in advance!

hypest commented:
> @hypest, would you be able to merge this today? I'll then go ahead with the other steps you listed. :) Thank you in advance!

Yeap, let's get this done! 😸

SiobhyB committed:
Merge remote-tracking branch 'upstream/trunk' into add-featured-badge

SiobhyB committed:
Update CHANGELOG.md

hypest reviewed (approved):
Still looking good to me, noticed the CHANGELOG.md changes as well and looks like the "Unreleased" section will need some updating to match the current (at the time of writing) native editor release v1.51.0, but that will be wrangled separately so, all good here 👍.
